calyx-backend/src/firrtl.rs

```diff
@@ -0,0 +1,181 @@
+
+    // Inputs and Outputs
+    let sig = comp.signature.borrow();
+    for (_idx, port_ref) in sig.ports.iter().enumerate() {
+        let port = port_ref.borrow();
+        let direction_string =
+        // NOTE: The signature port definitions are reversed inside the component.
+        match port.direction {
+            ir::Direction::Input => {"output"}
+            ir::Direction::Output => {"input"}
+            ir::Direction::Inout => {
+                panic!("Unexpected Inout port on Component: {}", port.name)
+            }
+        };
+        // FIXME: Hack to get clock declaration right. Should check for attribute name instead.
+        if port.name == "clk" {
+            writeln!(
+                f,
+                "{}{} {}: Clock",
+                SPACING.repeat(2),
+                direction_string,
+                port.name
+            )?;
+        } else {
+            writeln!(
+                f,
+                "{}{} {}: UInt<{}>",
+                SPACING.repeat(2),
+                direction_string,
+                port.name,
+                port.width
+            )?;
+        }
+    }
+
+    // Add a COMPONENT START: <name> anchor before any code in the component
+    writeln!(f, "{}; COMPONENT START: {}", SPACING.repeat(2), comp.name)?;
+
+    // TODO: Cells. NOTE: leaving this one for last
+
+    for asgn in &comp.continuous_assignments {
+        // TODO: guards
+        match asgn.guard.as_ref() {
+            ir::Guard::Or(_, _) => todo!(),
+            ir::Guard::And(_, _) => todo!(),
+            ir::Guard::Not(_) => todo!(),
+            ir::Guard::True => {
+                // Simple assignment with no guard
+                let _ = write_assignment(asgn, f);
+            }
+            ir::Guard::CompOp(_, _, _) => todo!(),
+            ir::Guard::Port(_) => {}
+            ir::Guard::Info(_) => todo!(),
+        }
+    }
+
+    // Add COMPONENT END: <name> anchor
+    writeln!(f, "{}; COMPONENT END: {}", SPACING.repeat(2), comp.name)?;
+
+    Ok(())
+}
+
+fn write_assignment<F: io::Write>(
+    asgn: &ir::Assignment<ir::Nothing>,
+    f: &mut F,
+) -> CalyxResult<()> {
+    let dest_port = asgn.dst.borrow();
+    let mut dest_string = SPACING.repeat(2);
+    match &dest_port.parent {
+        ir::PortParent::Cell(cell) => {
+            let parent_ref = cell.upgrade();
+            let parent = parent_ref.borrow();
+            match parent.prototype {
+                ir::CellType::ThisComponent => {
+                    dest_string.push_str(dest_port.name.as_ref());
+                }
+                _ => {
+                    let formatted = format!(
+                        "{}.{}",
+                        parent.name().as_ref(),
+                        dest_port.name.as_ref()
+                    );
+                    dest_string.push_str(&formatted);
+                }
+            }
+        }
+        _ => {
+            unreachable!()
```

Comment:
This should probably be a `todo` as it is possible to have assignments whose
ports have a group as a parent even though we aren't dealing with them yet. For
context, these ports are the go/done ports attached to groups. So
`my_group[done]` is a port named `done` whose parent is a group, rather than a cell.

Reply:
I was under the impression that I wouldn't need to cover any ports that have a group as a parent because this backend pass would take place after the compiler removes all groups... What do you think?

Reply:
oh yeah, you're totally right my bad. clearly got some wires crossed. You can add a message string to the macro call if you want but otherwise fine to leave as is. Sorry for the confusion!!

Reply:
Sounds good! Thanks :)